src/cargo/util/context/target.rs

```diff
@@ -53,6 +54,17 @@ pub(super) fn load_target_cfgs(
     let target: BTreeMap<String, TargetCfgConfig> = gctx.get("target")?;
     tracing::debug!("Got all targets {:#?}", target);
     for (key, cfg) in target {
+        if let Ok(Platform::Cfg(cfg_expr)) = key.parse() {
+            cfg_expr.walk_expr(|e| match e {
+                CfgExpr::True | CfgExpr::False => {
+                    if !gctx.cli_unstable().cfg_boolean_literals {
```

Comment:
Should we have a deprecation warning for `Name("true")` / `Name("false")` in existing versions, here and in `toml`?

If we do this, then no point in moving the if like I mentioned

Reply:
Strictly speaking, so long as a user doesn't set `--cfg false`, the use of `#[cfg(false)]` won't fundamentally change, so maybe we don't need a future-incompat for that?

Reply:
> Should we have a deprecation/future-incompat warning for `Name("true")` / `Name("false")` in existing versions, here and in toml?

We could, but I don't think it's worth it, we only have two users and I expect the RFC to become stable relatively quickly.

> Strictly speaking, so long as a user doesn't set `--cfg false`, the use of `#[cfg(false)]` won't fundamentally change

Technically that's not true for `rustc` as `#[cfg(false)]` was never legal before the RFC.

Reply:
Two users on github.  That isn't a complete search of all code.  I'm just hesitant about turning this into a hard error without any prior notice.

@weihanglo, thoughts?

Reply:
> Technically that's not true for rustc as #[cfg(false)] was never legal before the RFC.

If we treat it as a Cargo bug, we can literally ship this breaking change 😬.

Being conservative, I vote for warnings. We could give a “clear timeline” when it will become a hard error, and note that in our CHANGELOG.

Reply:
I wonder what other cfg value rustc disallows but Cargo is happy with. We may also fix them altogether.

Reply:
> If we treat it as a Cargo bug, we can literally ship this breaking change 😬.

That's my position, I would consider the Cargo parser to be in-sufficient and buggy, and just ship the breaking change ...

> Being conservative, I vote for warnings. We could give a “clear timeline” when it will become a hard error, and note that in our CHANGELOG.

... but that doesn't seems to be either of your position. 😄 

Which leads me to ask, what kind of warning you want? A future-incompatibility warning with no change in behaviour? A warning but with change in behaviour? Or a future-incompatibility warning but with unstable flag and feature to change the behaviour?

That last one would be my preferred one, as it would allow nightly testing, but it would be the most involved as currently the parsing is done via [`<CfgExpr as FromStr>::from_str(&str)`](https://doc.rust-lang.org/beta/nightly-rustc/cargo_platform/enum.CfgExpr.html#impl-FromStr-for-CfgExpr) which doesn't take any options, which will would need to configure the parser behaviour.

Reply:
I think an unstable-opt-in is fine to change from the future-incompat warning to the new behavior.

Reply:
We may also want to update the doc to note that some cfg values are not intended to use. Even if they are valid in `Cargo.toml` or `.cargo/config.toml`, they are unlikely to be picked by rustc.

There are two categories I can think of people might misuse:

* `[target.'cfg(true)'.dependencies]` being a way to activate additional features
  * The `[features]` table is the right approach of it.
* `target.'cfg(ture).rustflags` being a way to set/overwrite rustflags
  * `cargo --config` should be used instead, or `RUSTFLAGS` or `CARGO_BUILD_RUSTFLAGS`.

Reply:
Opened #14671 to address this thread *(as this PR is already quite large)*.